File: eye-gaze/api/python/wia_eye_gaze/dwell.py
"""
WIA Eye Gaze Standard - Dwell Controller

弘益人間 - 널리 인간을 이롭게
"""

import time
from typing import Optional, Dict, Set, Callable
import threading

from .tracker import IEyeTracker
from .types import (
    GazePoint,
    GazeTarget,
    DwellConfig,
    DwellState,
    Subscription,
)


DwellEventHandler = Callable[[GazeTarget, Optional[float]], None]


class DwellController:
    """
    Dwell Controller for gaze-based selection.

    Example:
        >>> dwell = DwellController(tracker=tracker, threshold=800)
        >>> dwell.register_target(GazeTarget(...))
        >>> dwell.on_dwell_complete(lambda t, _: print(f"Selected: {t.label}"))
        >>> dwell.start()
    """

    def __init__(
        self,
        tracker: IEyeTracker,
        threshold: int = 800,
        progress_interval: int = 50,
        visual_feedback: bool = True,
        cooldown_period: int = 500,
    ):
        self._tracker = tracker
        self._config = DwellConfig(
            threshold=threshold,
            progress_interval=progress_interval,
            visual_feedback=visual_feedback,
            cooldown_period=cooldown_period,
        )

        self._targets: Dict[str, GazeTarget] = {}
        self._dwell_state = DwellState(active=False, progress=0)
        self._subscription: Optional[Subscription] = None
        self._progress_timer: Optional[threading.Timer] = None
        self._cooldown_until = 0
        self._enabled = False

        self._on_dwell_start_handlers: Set[DwellEventHandler] = set()
        self._on_dwell_progress_handlers: Set[DwellEventHandler] = set()
        self._on_dwell_complete_handlers: Set[DwellEventHandler] = set()
        self._on_dwell_cancel_handlers: Set[DwellEventHandler] = set()

    def set_dwell_time(self, ms: int) -> None:
        """Set dwell selection time in milliseconds"""
        self._config.threshold = ms

    def get_dwell_time(self) -> int:
        """Get current dwell time"""
        return self._config.threshold

    def register_target(self, target: GazeTarget) -> None:
        """Register a target for dwell selection"""
        self._targets[target.element_id] = target

    def unregister_target(self, target_id: str) -> None:
        """Unregister a target"""
        self._targets.pop(target_id, None)

    def clear_targets(self) -> None:
        """Clear all targets"""
        self._targets.clear()
        self._cancel_dwell()

    def get_targets(self) -> list:
        """Get all registered targets"""
        return list(self._targets.values())

    def on_dwell_start(self, handler: DwellEventHandler) -> None:
        """Register handler for dwell start"""
        self._on_dwell_start_handlers.add(handler)

    def on_dwell_progress(self, handler: DwellEventHandler) -> None:
        """Register handler for dwell progress"""
        self._on_dwell_progress_handlers.add(handler)

    def on_dwell_complete(self, handler: DwellEventHandler) -> None:
        """Register handler for dwell complete"""
        self._on_dwell_complete_handlers.add(handler)

    def on_dwell_cancel(self, handler: DwellEventHandler) -> None:
        """Register handler for dwell cancel"""
        self._on_dwell_cancel_handlers.add(handler)

    def start(self) -> None:
        """Start dwell detection"""
        if self._enabled:
            return

        self._enabled = True
        self._subscription = self._tracker.subscribe(self._handle_gaze_data)

    def stop(self) -> None:
        """Stop dwell detection"""
        if not self._enabled:
            return

        self._enabled = False
        self._cancel_dwell()

        if self._subscription:
            self._tracker.unsubscribe(self._subscription)
            self._subscription = None

    def pause(self) -> None:
        """Temporarily pause dwell detection"""
        self._enabled = False
        self._cancel_dwell()

    def resume(self) -> None:
        """Resume dwell detection"""
        self._enabled = True

    def get_state(self) -> DwellState:
        """Get current dwell state"""
        return DwellState(
            active=self._dwell_state.active,
            progress=self._dwell_state.progress,
            target=self._dwell_state.target,
            start_time=self._dwell_state.start_time,
        )

    def dispose(self) -> None:
        """Dispose resources"""
        self.stop()
        self._targets.clear()
        self._on_dwell_start_handlers.clear()
        self._on_dwell_progress_handlers.clear()
        self._on_dwell_complete_handlers.clear()
        self._on_dwell_cancel_handlers.clear()

    def _handle_gaze_data(self, gaze: GazePoint) -> None:
        if not self._enabled or not gaze.valid:
            if self._dwell_state.active:
                self._cancel_dwell()
            return

        # Check cooldown
        if time.time() * 1000 < self._cooldown_until:
            return

        # Find target under gaze
        hit_target = self._find_target_at_point(gaze.x, gaze.y)

        if not hit_target:
            if self._dwell_state.active:
                self._cancel_dwell()
            return

        if not self._dwell_state.active:
            self._start_dwell(hit_target)
        elif self._dwell_state.target and self._dwell_state.target.element_id != hit_target.element_id:
            self._cancel_dwell()
            self._start_dwell(hit_target)

    def _find_target_at_point(self, x: float, y: float) -> Optional[GazeTarget]:
        for target in self._targets.values():
            if target.bounding_box.contains(x, y):
                return target
        return None

    def _start_dwell(self, target: GazeTarget) -> None:
        self._dwell_state = DwellState(
            active=True,
            target=target,
            start_time=int(time.time() * 1000),
            progress=0,
        )

        self._emit("start", target)
        self._start_progress_timer()

    def _start_progress_timer(self) -> None:
        def update():
            self._update_progress()
            if self._dwell_state.active:
                self._progress_timer = threading.Timer(
                    self._config.progress_interval / 1000,
                    update
                )
                self._progress_timer.daemon = True
                self._progress_timer.start()

        self._progress_timer = threading.Timer(
            self._config.progress_interval / 1000,
            update
        )
        self._progress_timer.daemon = True
        self._progress_timer.start()

    def _update_progress(self) -> None:
        if not self._dwell_state.active or not self._dwell_state.start_time or not self._dwell_state.target:
            return

        elapsed = int(time.time() * 1000) - self._dwell_state.start_time
        progress = min(1.0, elapsed / self._config.threshold)

        self._dwell_state.progress = progress
        self._emit("progress", self._dwell_state.target, progress)

        if progress >= 1.0:
            self._complete_dwell()

    def _complete_dwell(self) -> None:
        if not self._dwell_state.target:
            return

        target = self._dwell_state.target
        self._clear_progress_timer()
        self._dwell_state = DwellState(active=False, progress=0)
        self._cooldown_until = time.time() * 1000 + self._config.cooldown_period

        self._emit("complete", target, 1.0)

    def _cancel_dwell(self) -> None:
        if not self._dwell_state.active:
            return

        target = self._dwell_state.target
        self._clear_progress_timer()
        self._dwell_state = DwellState(active=False, progress=0)

        if target:
            self._emit("cancel", target)

    def _clear_progress_timer(self) -> None:
        if self._progress_timer:
            self._progress_timer.cancel()
            self._progress_timer = None

    def _emit(self, event: str, target: GazeTarget, progress: Optional[float] = None) -> None:
        handlers = {
            "start": self._on_dwell_start_handlers,
            "progress": self._on_dwell_progress_handlers,
            "complete": self._on_dwell_complete_handlers,
            "cancel": self._on_dwell_cancel_handlers,
        }.get(event, set())

        for handler in handlers:
            try:
                handler(target, progress)
            except Exception as e:
                print(f"Error in dwell {event} handler: {e}")


def create_dwell_controller(
    tracker: IEyeTracker,
    threshold: int = 800,
    **kwargs
) -> DwellController:
    """Create a dwell controller"""
    return DwellController(tracker=tracker, threshold=threshold, **kwargs)

File: eye-gaze/api/rust/src/dwell.rs
//! WIA Eye Gaze Standard - Dwell Controller
//!
//! 弘益人間 - 널리 인간을 이롭게

use std::collections::HashMap;
use std::sync::{Arc, Mutex};
use std::time::{Instant, SystemTime, UNIX_EPOCH};

use crate::types::*;

/// Dwell event handler type
pub type DwellEventHandler = Box<dyn Fn(&GazeTarget, Option<f64>) + Send + Sync>;

/// Dwell Controller for gaze-based selection
pub struct DwellController {
    config: DwellConfig,
    targets: Arc<Mutex<HashMap<String, GazeTarget>>>,
    state: Arc<Mutex<DwellState>>,
    cooldown_until: Arc<Mutex<u64>>,
    enabled: Arc<Mutex<bool>>,

    on_start_handlers: Arc<Mutex<Vec<DwellEventHandler>>>,
    on_progress_handlers: Arc<Mutex<Vec<DwellEventHandler>>>,
    on_complete_handlers: Arc<Mutex<Vec<DwellEventHandler>>>,
    on_cancel_handlers: Arc<Mutex<Vec<DwellEventHandler>>>,
}

impl DwellController {
    /// Create new dwell controller
    pub fn new(config: DwellConfig) -> Self {
        Self {
            config,
            targets: Arc::new(Mutex::new(HashMap::new())),
            state: Arc::new(Mutex::new(DwellState::default())),
            cooldown_until: Arc::new(Mutex::new(0)),
            enabled: Arc::new(Mutex::new(false)),
            on_start_handlers: Arc::new(Mutex::new(Vec::new())),
            on_progress_handlers: Arc::new(Mutex::new(Vec::new())),
            on_complete_handlers: Arc::new(Mutex::new(Vec::new())),
            on_cancel_handlers: Arc::new(Mutex::new(Vec::new())),
        }
    }

    /// Create with default config
    pub fn default() -> Self {
        Self::new(DwellConfig::default())
    }

    /// Set dwell threshold
    pub fn set_dwell_time(&mut self, ms: u64) {
        self.config.threshold_ms = ms;
    }

    /// Get dwell time
    pub fn get_dwell_time(&self) -> u64 {
        self.config.threshold_ms
    }

    /// Register target
    pub fn register_target(&self, target: GazeTarget) {
        let mut targets = self.targets.lock().unwrap();
        targets.insert(target.element_id.clone(), target);
    }

    /// Unregister target
    pub fn unregister_target(&self, target_id: &str) {
        let mut targets = self.targets.lock().unwrap();
        targets.remove(target_id);
    }

    /// Clear all targets
    pub fn clear_targets(&self) {
        let mut targets = self.targets.lock().unwrap();
        targets.clear();
        self.cancel_dwell();
    }

    /// Get targets
    pub fn get_targets(&self) -> Vec<GazeTarget> {
        let targets = self.targets.lock().unwrap();
        targets.values().cloned().collect()
    }

    /// Register dwell start handler
    pub fn on_dwell_start<F>(&self, handler: F)
    where
        F: Fn(&GazeTarget, Option<f64>) + Send + Sync + 'static,
    {
        let mut handlers = self.on_start_handlers.lock().unwrap();
        handlers.push(Box::new(handler));
    }

    /// Register dwell progress handler
    pub fn on_dwell_progress<F>(&self, handler: F)
    where
        F: Fn(&GazeTarget, Option<f64>) + Send + Sync + 'static,
    {
        let mut handlers = self.on_progress_handlers.lock().unwrap();
        handlers.push(Box::new(handler));
    }

    /// Register dwell complete handler
    pub fn on_dwell_complete<F>(&self, handler: F)
    where
        F: Fn(&GazeTarget, Option<f64>) + Send + Sync + 'static,
    {
        let mut handlers = self.on_complete_handlers.lock().unwrap();
        handlers.push(Box::new(handler));
    }

    /// Register dwell cancel handler
    pub fn on_dwell_cancel<F>(&self, handler: F)
    where
        F: Fn(&GazeTarget, Option<f64>) + Send + Sync + 'static,
    {
        let mut handlers = self.on_cancel_handlers.lock().unwrap();
        handlers.push(Box::new(handler));
    }

    /// Start dwell detection
    pub fn start(&self) {
        let mut enabled = self.enabled.lock().unwrap();
        *enabled = true;
    }

    /// Stop dwell detection
    pub fn stop(&self) {
        let mut enabled = self.enabled.lock().unwrap();
        *enabled = false;
        self.cancel_dwell();
    }

    /// Pause dwell detection
    pub fn pause(&self) {
        let mut enabled = self.enabled.lock().unwrap();
        *enabled = false;
        self.cancel_dwell();
    }

    /// Resume dwell detection
    pub fn resume(&self) {
        let mut enabled = self.enabled.lock().unwrap();
        *enabled = true;
    }

    /// Get current state
    pub fn get_state(&self) -> DwellState {
        self.state.lock().unwrap().clone()
    }

    /// Process gaze data
    pub fn process_gaze(&self, gaze: &GazePoint) {
        if !*self.enabled.lock().unwrap() || !gaze.valid {
            if self.state.lock().unwrap().active {
                self.cancel_dwell();
            }
            return;
        }

        // Check cooldown
        let now = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_millis() as u64;

        if now < *self.cooldown_until.lock().unwrap() {
            return;
        }

        // Find hit target
        let hit_target = self.find_target_at(gaze.x, gaze.y);

        match hit_target {
            None => {
                if self.state.lock().unwrap().active {
                    self.cancel_dwell();
                }
            }
            Some(target) => {
                let current_state = self.state.lock().unwrap().clone();

                if !current_state.active {
                    self.start_dwell(target);
                } else if let Some(ref current_target) = current_state.target {
                    if current_target.element_id != target.element_id {
                        self.cancel_dwell();
                        self.start_dwell(target);
                    } else {
                        self.update_progress();
                    }
                }
            }
        }
    }

    fn find_target_at(&self, x: f64, y: f64) -> Option<GazeTarget> {
        let targets = self.targets.lock().unwrap();
        targets
            .values()
            .find(|t| t.bounding_box.contains(x, y))
            .cloned()
    }

    fn start_dwell(&self, target: GazeTarget) {
        let now = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_millis() as u64;

        {
            let mut state = self.state.lock().unwrap();
            state.active = true;
            state.target = Some(target.clone());
            state.start_time = Some(now);
            state.progress = 0.0;
        }

        self.emit_event(&self.on_start_handlers, &target, None);
    }

    fn update_progress(&self) {
        let now = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_millis() as u64;

        let (target, progress, complete) = {
            let mut state = self.state.lock().unwrap();

            if !state.active || state.target.is_none() || state.start_time.is_none() {
                return;
            }

            let elapsed = now - state.start_time.unwrap();
            let progress = (elapsed as f64 / self.config.threshold_ms as f64).min(1.0);
            state.progress = progress;

            (state.target.clone().unwrap(), progress, progress >= 1.0)
        };

        self.emit_event(&self.on_progress_handlers, &target, Some(progress));

        if complete {
            self.complete_dwell();
        }
    }

    fn complete_dwell(&self) {
        let target = {
            let mut state = self.state.lock().unwrap();
            let target = state.target.clone();
            state.active = false;
            state.progress = 0.0;
            state.target = None;
            state.start_time = None;
            target
        };

        if let Some(target) = target {
            let now = SystemTime::now()
                .duration_since(UNIX_EPOCH)
                .unwrap()
                .as_millis() as u64;

            *self.cooldown_until.lock().unwrap() = now + self.config.cooldown_period_ms;
            self.emit_event(&self.on_complete_handlers, &target, Some(1.0));
        }
    }

    fn cancel_dwell(&self) {
        let target = {
            let mut state = self.state.lock().unwrap();
            if !state.active {
                return;
            }
            let target = state.target.clone();
            state.active = false;
            state.progress = 0.0;
            state.target = None;
            state.start_time = None;
            target
        };

        if let Some(target) = target {
            self.emit_event(&self.on_cancel_handlers, &target, None);
        }
    }

    fn emit_event(
        &self,
        handlers: &Arc<Mutex<Vec<DwellEventHandler>>>,
        target: &GazeTarget,
        progress: Option<f64>,
    ) {
        let handlers = handlers.lock().unwrap();
        for handler in handlers.iter() {
            handler(target, progress);
        }
    }
}
